src/runtime/providers/aggregate.ts

```diff
@@ -176,8 +176,79 @@ export class Max extends Aggregate {
   }
 }
 
+export class Join extends Aggregate {
+  static AttributeMapping = {
+    "value": 0,
+    "given": 1,
+    "per": 2,
+    "ordinal": 3,
+    "separator": 4,
+  }
+
+  ordinal : any;
+  separator : any;
+
+  constructor(id: string, args: any[], returns: any[]) {
+    super(id, args, returns);
+    this.ordinal = args[3]
+    this.separator = args[4]
+  }
+
+  aggregate(rows: any[]) {
+    let groupKeys = [];
+    let groups = {};
+    for(let row of rows) {
+      resolve(this.projectionVars, row, this.resolvedProjection)
+      resolve(this.groupVars, row, this.resolvedGroup)
+      let group = this.resolvedAggregate.group;
+      let projection = this.resolvedAggregate.projection;
+      let value = toValue(this.value, row);
+      let ordinal = toValue(this.ordinal, row);
+      let separator = toValue(this.separator, row);
+
+      if (separator === undefined) separator = "";
+
+      let groupKey = "[]";
+      if(group.length !== 0) {
+        groupKey = JSON.stringify(group);
+      }
+      let groupValues = groups[groupKey];
+      if(groupValues === undefined) {
+        groupKeys.push(groupKey);
+        groupValues = groups[groupKey] = {result:[]};
+      }
+      let projectionKey = JSON.stringify(projection);
+      if(groupValues[projectionKey] === undefined) {
+        groupValues[projectionKey] = true;
+        groupValues.result.push({value: value, ordinal:ordinal, separator:separator})
+      }
+    }
+
+    for (let g in groups) {
+        let s = groups[g].result.sort((a, b) => 
+                                      {if (a.ordinal > b.ordinal) return 1;
+                                       if (a.ordinal === b.ordinal) return 0;
+                                       return -1;})
+        let len = s.length
+        let result = ""
+        for (var i=0; i<len; ++i) {
+            // this means that the sep assocated with a value
+            // is the one which occurs before the value
+            if (i != 0) result += s[i].separator
```

Comment:
i know i wrote a comment here. we should decide if its the right thing, but the sep (with) precedes  the text (token), so since we lop off the first, only the interstitial are left. i haven't been able to come up with a clever mapping that lets us specify all the options, but but haven't given it too much thought